presto-main/src/main/java/com/facebook/presto/operator/scalar/DateTimeFunctions.java

```diff
@@ -83,8 +84,11 @@ private DateTimeFunctions() {}
     @SqlType(StandardTypes.DATE)
     public static long currentDate(ConnectorSession session)
     {
-        long millis = getChronology(session.getTimeZoneKey()).dayOfMonth().roundFloor(session.getStartTime());
-        return MILLISECONDS.toDays(millis);
+        ISOChronology chronology = getChronology(session.getTimeZoneKey());
+
+        DateTime currentDateTime = new DateTime(System.currentTimeMillis(), chronology).withTimeAtStartOfDay();
+        DateTime baseDateTime = new DateTime(1970, 1, 1, 0, 0, chronology).withTimeAtStartOfDay();
+        return Days.daysBetween(baseDateTime, currentDateTime).getDays();
```

Comment:
This code is a lot more expensive than the previous code because it is constructing two `DateTime` objects.  Can you dig into the underlying logic of these methods and rewrite this code to operate only on primitive types?